VLAN support

jukkar opened the pull request:
This set provides VLAN support for ethernet networks. Works when connecting to Linux host with both IPv4 and IPv6 and with multiple VLAN tags.
To enable VLAN support one needs to set **CONFIG_NET_VLAN=y** in config file.
To enable VLAN traffic, currently this needs to be enabled at runtime in net-shell with "_**net vlan add 100 0**_" command. This will add VLAN tag 100 to first network interface (index 0). Multiple tags can be enabled simultaneously, one just needs to set **CONFIG_NET_VLAN_COUNT=xxx** to proper value (default is 1).  The sample app will automatically enable configured VLANs.

This PR builds on top of following PRs:

* PR #6601 Checksum offloading
* PR #6076 Traffic classification

jukkar committed:
net: if: Separate IP address configuration from net_if

Move IP address settings from net_if to separate structs.
This is needed for VLAN support.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: eth: mcux: Enabling VLAN

This enables / fixes VLAN support in mcux ethernet driver.

The commit contains these changes for enabling VLAN:
* Increase the size of the ethernet frame if VLAN is enabled.
* Enable VLAN in chip if VLAN is enabled
* If VLAN is enabled, then the iface in context struct should
  not be used directly as there can be multiple VLAN iface
  related to this physical device.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
This apparently addresses https://github.com/zephyrproject-rtos/zephyr/issues/3234

tbursztyka reviewed (changes requested):
Nack on the current approach.

It's introducing an L2 and bearer specific attribute to the net core part. Then we have vlan_tag as a parameter everywhere, and that's really bad.
Plus, having a fixed size of net_if_config on each iface means we will use too much memory in some use case: one iface with 2 vlan, another one with 5, etc...

DAD, mcast etc... all of this don't use the vlan_tag as is. What matters here is the ip configuration only and its relationship with iface. And that's where we should address the changes.

We could split net_if in 2: 
- net_if_dev (the one that relates to an actual device):
   the same as what net_if is right now.
- net_if_virt (a "virtual" iface, which in this case would be instantiated to hold vlan ip config stuff):
  that one would hold one net_if_config and a pointer to       the net_if_dev it is attached to.

net_if being then just a generic place-holder of a union on these 2 structs, with an attribute that can be used to differentiate those.

Something like that. It's more work, but at least it won't spread vlan stuff everywhere.

The core part of vlan itself should go in L2 (either tight to ethernet, or more generic as - maybe - future tech support would require it, such as ATM based devs etc...).

jukkar committed:
samples: net: vlan: Simple app for setting virtual lan settings

This application just enables VLAN tag for ethernet interface.
Set CONFIG_SAMPLE_VLAN_TAG option to define the desired VLAN tag.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
> Plus, having a fixed size of net_if_config on each iface means we will use too much memory in some use case: one iface with 2 vlan, another one with 5, etc...

Yep, agreed on this one. I did not try to separate IP config from net_if yet, there would have been too much changes. We could have the config set separately and only store pointer to it in net_if in order to save space when some functionality is not needed.

> 
> We could split net_if in 2:
> 
>     net_if_dev (the one that relates to an actual device):
>     the same as what net_if is right now.
>     net_if_virt (a "virtual" iface, which in this case would be instantiated to hold vlan ip config stuff):
>     that one would hold one net_if_config and a pointer to the net_if_dev it is attached to.

The problem is that currently in the code we have "struct net_if *" in many places. We could do it so that current net_if is renamed to net_if_dev, and then create new net_if that would contain IP configuration etc that is needed for the interface. We probably do not need this net_if_virt at all.

> 
> net_if being then just a generic place-holder of a union on these 2 structs, with an attribute that can be used to differentiate those.

I am not sure this is needed but lets see how this goes.

Anyway, this set enables a way to have VLAN support which was the intention. The changes in this set are quite simple actually, we just need to generalize this a bit more. The purpose of this set was to get some comments and now they are got.

I will send a new version of this set to fix various bugs, no need to review it at this time.

jukkar committed:
net: shell: Add VLAN support

Add commands to add, remove or get information about VLANs
attached to network interfaces.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version uploaded that adds VLAN support only to ethernet L2 driver.
This one is still missing:
- vlan priority
- IP address optimizations (now uses too much memory)
These two are work in progress.

dbkinder commented on samples/net/vlan/README.rst:
```diff
@@ -0,0 +1,47 @@
+usage scenario is to run vlan application inside QEMU. This is
+described in :ref:`networking_with_qemu`.
+
+Build VLAN sample application like this:
```
Follow these steps to build the VLAN sample application:

dbkinder commented on samples/net/vlan/README.rst:
```diff
@@ -0,0 +1,47 @@
+   :goals: build
+   :compact:
+
+The default configuration file prj.conf will create two virtual lan networks
```
s/will create/creates/
s/lan/LAN/

dbkinder commented on samples/net/vlan/README.rst:
```diff
@@ -0,0 +1,47 @@
+=====================
+
+The :file:`samples/net/vlan/vlan-setup-linux.sh` provides a script that
+can be executed in Linux host. It creates two VLANs in Linux host and
```
s/in/on the/  (twice)

dbkinder reviewed (changes requested):
suggested edits

jukkar committed:
net: if: vlan: Add virtual lan support

This allows creation of virtual lan (VLAN) networks. VLAN support is
only available for ethernet network technology.

Fixes #3234

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Make interface IP configuration more flexible

Instead of always allocating both IPv6 and IPv4 address information
to every network interface, allow more fine grained address
configuration. So it is possible to have IPv6 or IPv4 only network
interfaces.

This commit introduces two new config options:
CONFIG_NET_IF_MAX_IPV4_COUNT and CONFIG_NET_IF_MAX_IPV6_COUNT
which tell how many IP address information structs are allocated
statically. At runtime when network interface is setup, it is then
possible to attach this IP address info struct to a specific
network interface. This can save considerable amount of memory
as the IP address information struct can be quite large (depends
on how many IP addresses user configures in the system).

Note that the value of CONFIG_NET_IF_MAX_IPV4_COUNT and
CONFIG_NET_IF_MAX_IPV6_COUNT should reflect the estimated number of
network interfaces in the system. So if if CONFIG_NET_IF_MAX_IPV6_COUNT
is set to 1 and there are two network interfaces that need IPv6
addresses, then the system will not be able to setup IPv6 addresses to
the second network interface in this case. This scenario might be
just fine if the second network interface is IPv4 only. The net_if.c
will print a warning during startup if mismatch about the counts and
the actual number of network interface is detected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated new version:
- documentation fixed
- the code passes now sanitycheck
- optimization of IP addresses in network interface (can save lot of memory when having multiple network interfaces)

jukkar commented:
New version that fixes sanitycheck error. I did not see this one when manually running sanitycheck but shippable run sanitycheck gives some errors.

jukkar commented:
Documentation fix to satisfy sanitycheck.

jukkar commented:
New version that fixes merge conflict.

dbkinder commented on samples/net/vlan/README.rst:
```diff
@@ -0,0 +1,58 @@
+
+The default configuration file prj.conf creates two virtual LAN networks
+with these settings:
+- VLAN tag 100: IPv4 198.51.100.1 and IPv6 2001:db8:100::1
```
add a blank line before a list (otherwise the line looks like a continuation of the sentence)

dbkinder reviewed (changes requested):
one last formatting fix

tbursztyka commented on include/net/net_pkt.h:
```diff
@@ -128,6 +128,11 @@ struct net_pkt {
+
+#if defined(CONFIG_NET_VLAN)
+	u16_t vlan_tag;
+#endif /* CONFIG_NET_VLAN */
```
Why net_pkt would need to store that info?

jukkar replied:
I noticed that iface variable that is feed to L2 ethernet.c is not always the proper value for a specific VLAN tag. Having this value set in net_pkt is also useful as figuring it out from iface is time consuming as iface does not actually know anything about VLANs.

tbursztyka replied:
ok

tbursztyka commented on subsys/net/ip/net_shell.c:
```diff
@@ -161,6 +161,9 @@ static void iface_cb(struct net_if *iface, void *user_data)
 #if defined(CONFIG_NET_IPV6)
 	struct net_if_ipv6_prefix *prefix;
 	struct net_if_router *router;
+#endif
```
don't you have a dedicated patch for shell + vlan? that part should go there.

jukkar replied:
That patch is for creating "net vlan" command, the changes here just print vlan information when user gives "net iface" command.

tbursztyka reviewed (changes requested):
Github was unable to load the large diffs for me...

Anyway, in arp and ethernet, could create dedicated function for setting up the ptype, hdr so it would unbloat a bit the functions. It's full of ifdefs and if/else there. It would be more readable to isolate things a bit.

jukkar committed:
net: if: Use DEVICE_NAME_GET() instead of fixed string

The DEVICE_NAME_GET() macro should be used instead of fixed
string when creating a device pointer in net_if_dev structure.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented on include/net/net_ip.h:
```diff
@@ -349,9 +349,10 @@ static inline bool net_is_ipv6_addr_mcast(const struct in6_addr *addr)
-extern struct net_if_addr *net_if_ipv6_addr_lookup(const struct in6_addr *addr,
-						   struct net_if **iface);
+extern struct net_if_addr *net_if_ipv6_addr_lookup(
+	const struct in6_addr *addr, struct net_if **iface);
```
Yes, this change is unnecessary, will fix.

jukkar commented:
> It's full of ifdefs and if/else there. It would be more readable to isolate things a bit.

That is unfortunate but I am not sure how to isolate things in arp.c, I will investigate if that is possible.

jukkar committed:
net: if: Add helper to select src interface for a IPv4 dst addr

If we have multiple network interfaces and we want to send
a IPv4 network packet to certain destination, then this new
helper can be used to figure out what network interface to use.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Use correct network interface for IPv4 ping

Instead of always using default interface, use the IPv4 target
address to select the correct network interface when sending
IPv4 ping request.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated according to comments. Also fixed couple of issues I noticed while testing VLAN support.

jukkar commented:
Updated new version as newly merged openthread code was using IP address configs that this PR changes, thus there was various compile errors.

jukkar committed:
tests: net: vlan: Add VLAN tests

Add tests to verify the bit manipulation functions in net/vlan.h
file. Also check that we can check if given network interface has
VLAN enabled or not. Verify also that received network packet
contains correct VLAN tag.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version uploaded:

* rebased because of merge conflicts
* VLAN priority support added
* unit tests for VLAN bit field functions

jukkar committed:
net: context: Add PRIORITY option support

Add context option support and implement PRIORITY option that
can be used to classify the network traffic to different trafic
classes according to said priority value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: core: Add array support to stack info helpers

Allow caller to create array of thread stacks using
NET_STACK_ARRAY_DEFINE() macro. This allows more debug information
to be printed by "net stacks" command.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Add initial TX and RX traffic class support

With this commit it is possible to add priority to sent or received
network packets. So user is able to send or receive higher priority
packets faster than lower level packets.
The traffic class support is activated by CONFIG_NET_TC_COUNT option.
The TC support uses work queues to separate the traffic. The
priority of the work queue thread specifies the ordering of the
network traffic. Each work queue thread handles traffic to one specific
work queue. Note that you should not enable traffic classes unless
you really need them by your application. Each TC thread needs
stack so this feature requires more memory.

It is possible to disable transmit traffic class support and keep the
receive traffic class support, or vice versa. If both RX and TX traffic
classes are enabled, then both will use the same number of queues
defined by CONFIG_NET_TC_COUNT option.

Fixes #6588

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: stats: Do not calculate total packet length many times

Calculate network packet total length once and then use that
value instead of calculating it many times in a row.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print TX and RX threads stacks with more info

The "net stacks" command was printing TX or RX thread values so
that it was not possible to use the values in debugging easily.
Add traffic class value when printing the info so that it is
easy to see what TX or RX queue is doing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: stats: Add traffic class statistics

Add statistics for number of packets and bytes to each traffic
class. Print this information in net-shell.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: l2: ethernet:  Add priority to sent ethernet VLAN header

When sending packet add the VLAN priority to the ethernet header.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: Add traffic class sample application

This demostrates how to classify the network traffic when sending
data. The application will create similar functionality as
echo-client so user can use echo-server running in remote host to
test this sample application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version that fixes the sanity errors.

jukkar committed:
net: app: Select local IP address properly if multiple interfaces

If we have multiple network interface (like in VLAN), then we need
to select the proper local interface based on destination address.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed the compilation if IPv4 is not enabled.

tbursztyka commented on subsys/net/ip/net_if.c:
```diff
@@ -190,11 +199,12 @@ static void net_if_process_events(struct k_poll_event *event, int ev_count)
-			iface = CONTAINER_OF(event->fifo, struct net_if,
-					     tx_queue);
-			net_if_tx(iface);
+			net_if_foreach(iface_tx_cb, if_dev);
```
Ok, this is temporary as new PR will change that entirely (TC support etc...) So it won't need to loop on each iface etc...

tbursztyka reviewed (changes requested):
Merge the 3 mcux patch in one (and call the patch "Enabling VLAN in mcux" or something like that). No need to be incremental here, and maybe you want to order the patch in the list like right before the test patch maybe.

jukkar commented:
Updated the PR according to comments.

dbkinder reviewed (approved):
thanks

jukkar commented:
Updated because of changes in master. New net_if_carrier_down() func caused compile error with this PR.

jukkar commented:
Uploaded because of merge conflict

jukkar committed:
samples: net: zperf: Use all the RAM and ROM for Quark SE dev

As the config file disables ARC we can use its RAM and ROM in
x86 side.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: wpan-serial: Use all the RAM and ROM for Quark SE dev

As the config file disables ARC we can use its RAM and ROM in x86 side.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Added more unit tests.

jukkar committed:
net: l2: Remove l2_data section start and end pointers

As the l2_data section might contain different size context elements
like "struct ethernet_context" for Ethernet and "void *" for
Dummy L2, remove the __net_l2_start and __net_l2_end variables so
that user does not accidentally try to use them as that would not work.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version that is now on top of traffic class patches. This was needed as I could not get the tx_queue handling work reliably and the traffic class PR got rid of that part of the code in this PR.

jukkar commented:
New version with sanitychecker fixes.

jukkar committed:
drivers: eth: native_posix: Add VLAN support

Support also virtual LAN (VLAN) with native_posix ethernet driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: slip: Add multiple VLAN interface support

Allow multiple VLANs to be configured for SLIP.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: Add unit test for traffic class support

Test that we can set the priority of the network packet and
that high priority packets are sent before low priority ones.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed slip support.

jukkar committed:
net: Add function to convert VLAN priority to packet priority

Currently the VLAN priority is the same as packet priority but
if such conversion is needed, then this function can be used
for such conversion.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: net: mcux: Use VLAN priority to set RX packet priority

Set the received network packet priority according to VLAN priority.
Currently this mapping is 1:1 but can be changed if needed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Rebase against traffic class pr.

jukkar committed:
tests: net: Increase the size of main stack

This avoids this crashing in qemu_x86

***** Stack Check Fail! *****
Current thread ID = 0x004015e0
Faulting segment:address = 0x0008:0x00003593
eax: 0x0041e930, ebx: 0x00000000, ecx: 0x004170bc, edx: 0x0000dff8
esi: 0xffffffff, edi: 0x0041e930, ebp: 0x00417010, esp: 0x00416ffc
eflags: 0x216
Fatal fault in essential thread! Spinning...
Terminate emulator due to fatal kernel error

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Rebase against master. Fixed issues reported by sanitycheck.

jukkar committed:
tests: net: Network packet checksum offloading tests

Test that if the checksum offloading is enabled, then we do not
calculate the checksum. Also the normal case, where offloading is
disabled and we need to calculate the checksum, is tested.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Rebase against master.

jukkar committed:
net: Support network packet checksum calc offloading

Create infrastructure that allows ethernet device driver to tell
if it supports network packet checksum offloading. This applies only
to IPv4, UDP or TCP checksums. The driver can enable/disable checksum
offloading separately for Tx and Rx network packets.

Fixes #2987

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Skip network packet checksum calc if it is offloaded

If the device (ethernet in this case) can calculate the network
packet checksum for IPv4, UDP or TCP, then do not calculate the
corresponding checksum by the stack itself.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
syslog: net: Fix multiple network interface selection for IPv4

In IPv4 we need to select the network interface, where the packet
is to be sent, using the IPv4 address instead of default network
interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Add network interface up/down command

User is able to take a network interface down or bring it up.
The command syntax is "net iface [up|down] [index]"

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
* Rebased against master
* Fixed checksum_offloading tests because of VLAN changes
* Added "net iface up|down" command to net-shell

pfalcon commented:
> Added "net iface up|down" command to net-shell

Would it make sense to submit that as a separate PR for quick approval? (I for example didn't yet have a chance to go thru all 32 commits of this PR.)

jukkar commented:
rebase against master

jukkar commented:
Removing this at this time as this seems to confuse people too much. I will re-send this, after PRs that depend on it, have been merged.
